lib/puppetserver/ca/certificate_authority.rb

```diff
@@ -0,0 +1,210 @@
+        results = @client.with_connection(url) do |connection|
+          certnames.map do |certname|
+            url.resource_name = certname
+            result = connection.put(body, url)
+            process_results(type, certname, result)
+          end
+        end
+
+        results.all?
+      end
+
+      # logs the action and returns true/false for success
+      def process_results(type, certname, result)
+        case type
+        when :sign
+          case result.code
+          when '204'
+            @logger.inform "Successfully signed certificate request for #{certname}"
+            return true
+          else
+            @logger.err 'Error:'
+            @logger.err "    When signing request submitted for #{certname}:"
+            @logger.err "      code: #{result.code}"
+            @logger.err "      body: #{result.body.to_s}" if result.body
+            return false
+          end
+        when :revoke
+          case result.code
+          when '200', '204'
+            @logger.inform "Revoked certificate for #{certname}"
+            return true
+          when '404'
+            @logger.err 'Error:'
+            @logger.err "    Could not find certificate for #{certname}"
+            return false
+          else
+            @logger.err 'Error:'
+            @logger.err "    When revoking #{certname} received:"
+            @logger.err "      code: #{result.code}"
+            @logger.err "      body: #{result.body.to_s}" if result.body
+            return false
+          end
+        when :submit
+          case result.code
+          when '200', '204'
+            @logger.inform "Successfully submitted certificate request for #{certname}"
+            return true
+          else
+            @logger.err 'Error:'
+            @logger.err "    When certificate request submitted for #{certname}:"
+            @logger.err "      code: #{result.code}"
+            @logger.err "      body: #{result.body.to_s}" if result.body
+            return false
+          end
+        end
+      end
+
+      # Make an HTTP request to CA to clean the named certificates
+      # @param certnames [Array] the name of the certificate(s) to have cleaned
+      # @return [Boolean] whether all certificate cleaning and revocation was successful
+      def clean_certs(certnames)
+        url = make_ca_url('certificate_status')
+
+        results = @client.with_connection(url) do |connection|
+          certnames.map do |certname|
+            url.resource_name = certname
+            revoke_result = connection.put(REVOKE_BODY, url)
+            revoked = check_revocation(certname, revoke_result)
+
+            cleaned = nil
+            unless revoked == :error
+              clean_result = connection.delete(url)
+              cleaned = check_clean(certname, clean_result)
+            end
+
+            cleaned == :success && [:success, :not_found].include?(revoked)
+          end
+        end
+
+        return results.all?
```

Comment:
Agree with Justin, I think we should move the verification into a block passed to `all`. I find this dangling call here at the end hard to read.